packages/web/server/queries/complex/pools/bonding.ts

```diff
@@ -0,0 +1,307 @@
+import { CoinPretty, PricePretty, RatePretty } from "@keplr-wallet/unit";
+import dayjs from "dayjs";
+import { Duration } from "dayjs/plugin/duration";
+
+import { querySyntheticLockupsByLockId } from "../../osmosis/lockup";
+import {
+  querySuperfluidDelegations,
+  querySuperfluidUnelegations,
+} from "../../osmosis/superfluid";
+import { calcSumCoinsValue } from "../assets";
+import { DEFAULT_VS_CURRENCY } from "../assets/config";
+import { getUserLocks } from "../osmosis";
+import { getValidatorInfo } from "../staking/validator";
+import {
+  getActiveGauges,
+  getCachedPoolIncentivesMap,
+  getIncentivizedPools,
+  getLockableDurations,
+} from "./incentives";
+import { getGammShareUnderlyingCoins, makeGammShareCurrency } from "./share";
+import { getSuperfluidPoolIds } from "./superfluid";
+
+export type UserDurationSuperfluidDelegation = {
+  delegated?: CoinPretty;
+  undelegating?: CoinPretty;
+
+  // validator info
+  commission?: RatePretty;
+  validatorMoniker?: string;
+  validatorLogoUrl?: string;
+};
+
+/** Bond duration that corresponds to locked pool shares. */
+export type BondDuration = {
+  duration: Duration;
+  /** Bondable if there's any active gauges for this duration. */
+  bondable: boolean;
+  /** User locked shares. */
+  userShares: CoinPretty;
+  userLockedShareValue: PricePretty;
+  userLocks: {
+    lockId: string;
+    /** AKA is superfluid lock */
+    isSynthetic: boolean;
+  }[];
+  userUnlockingShares?: { shares: CoinPretty; endTime?: Date };
+  aggregateApr: RatePretty;
+  swapFeeApr: RatePretty;
+  incentivesBreakdown: {
+    apr: RatePretty;
+    type: "osmosis" | "boost";
+  }[];
+  /** Both `delegated` and `undelegating` will be `undefined` if the user may "Go superfluid". */
+  superfluid?: {
+    /** Duration users can bond to for superfluid participation. Assumed to be longest duration on lock durations chain param. */
+    duration: Duration;
+    apr: RatePretty;
+  } & UserDurationSuperfluidDelegation;
+};
+
+export async function getSharePoolBondDurations(
+  poolId: string,
+  bech32Address?: string
+): Promise<BondDuration[]> {
+  // const sharePool = await getSharePool(poolId);
+  const activeGauges = await getActiveGauges();
+  const poolGauges = activeGauges.filter(
+    (gauge) =>
+      gauge.distribute_to.lock_query_type === "ByDuration" &&
+      gauge.distribute_to.denom.endsWith(poolId)
+  );
+  const poolIncentives = (await getCachedPoolIncentivesMap()).get(poolId);
+  const superfluidPoolIds = await getSuperfluidPoolIds();
```

Comment:
Can we use a Promise.all to solve all these queries? So we can reduce waterfall problems?

Reply:
Done, see https://github.com/osmosis-labs/osmosis-frontend/pull/2885/commits/42ac427c8dc1afa13a59575b8fa6cac1abc68f90